### DISCUSSION TITLE
Tonemapping HDR to SDR with libplacebo

### DISCUSSION THREAD

**Discussion by ggarra13** [haasn/libplacebo#294] [Q&A]
I am investigating using libplacebo for tonemapping HDR to SDR, as it seems the most capable library.
However, I need to integrate it with a custom pipeline using OpenColorIO and OpenGL.
Is there any way to output the shader code created by libplacebo into its own function (or to see what it does behind the scenes for debugging), like in OpenColorIO, so that my OpenGL wrapper code calls (in pseudo code):
```
// to convert from ocio's input space to linear space
outColor = ocioICS(outColor);

// do further color corrections on outColor like exposure/tint/contrast/etc..

// Operate libplacebo on linear space
if (toneMapEnabled)
{
   outColor = libplacebo_tonemapHDR(outColor);
}
// Take the resulting HDR tonemap colors to the OCIO Display (Usually Rec709 or sRGB)
outColor = ocioDisplayView(outColor);
```
?

**Comment by haasn** [comment]
Yes, this is supported. If you want to fully debug/dump the generated shaders, the most flexible way to do this is:

1. Create a dummy `pl_gpu` using [<libplacebo/dummy.h>](https://github.com/haasn/libplacebo/blob/master/src/include/libplacebo/dummy.h)
2. Allocate a `pl_shader` object using [pl_shader_alloc](https://github.com/haasn/libplacebo/blob/master/src/include/libplacebo/shaders.h#L70), passing the `pl_gpu` you created in step 1
3. Call [pl_shader_color_map_ex](https://github.com/haasn/libplacebo/blob/master/src/include/libplacebo/shaders/colorspace.h#L359) on this shader object.
4. Finalize the shader with [pl_shader_finalize](https://github.com/haasn/libplacebo/blob/master/src/include/libplacebo/shaders.h#L247) when you are done.

The resulting `pl_shader_res` contains the GLSL function text in `const char *glsl`, callable with `const char *name`, alongside a list of all descriptors you will need to make available (with the given names). You can unpack the texture data with [pl_buf_dummy_data](https://github.com/haasn/libplacebo/blob/master/src/include/libplacebo/dummy.h#L107).

Of course, for best results, you'll also want to run a separate HDR signal analysis pass using [pl_shader_detect_peak](https://github.com/haasn/libplacebo/blob/master/src/include/libplacebo/shaders/colorspace.h#L167), and optionally a HDR feature extraction pass.

Edit: Of course, I should point out, that if you're already using OpenGL, and you're comfortable with having libplacebo call into your GL context directly, you can vastly simplify the process of mapping all textures etc by just using a `pl_gpu` object created by `pl_opengl_create`. Then you can just `pl_opengl_unwrap` them and get your already-uploaded OpenGL texture objects ready to be used. This is what VLC does internally to do tone-mapping via libplacebo.

**Comment by ggarra13** [comment]
Can you explain to me how to get and pass the LUT textures into opengl from libplacebo when the pl_shader_color_map_ex()'s pl_color_map_args' **state** is set and the tone mapping function is **pl_tone_map_spline** or similar?  It seems it creates some LUTs, but I cannot figure how to add them to my code or how to pass them on to the GL backend.

If you have sample or test code that you can point me to, it would be ideal.  I apologyze for my dumb questions, but I am still new to libplacebo's great API.  You should be really proud of yourself @haasn!  You created a marvelous library!

**Comment by ggarra13** [comment]
Never mind.  I figured it out by myself.  I had to use the pl_tex* functions to extract the LUT data.

**Comment by wyup** [comment]
I have some questions:

Why does libplacebo bt2446a seem to use MaxDML instead of MaxCLL for default tone mapping upper limit? In my opinion it should use MaxCLL because that is the highest luminance limit of the content, MaxDML is not used so it reduces the effective tone mapping performance for a range that is not used.

Why does libplacebo use 203.0 nits by default for max limit for SDR tonemapping instead of standard 100 nits? Maybe it works better, but I have the doubt... Is it because it takes 0.2023 nits as minimum limit for SDR aswell? 203/0,2023 = ~ 100 units of dynamic range..

What default values does `color_map_preset=high_quality` change?

Thank you for this wonderful library.

**Reply by haasn** [reply]
> Why does libplacebo bt2446a seem to use MaxDML instead of MaxCLL for default tone mapping upper limit? 

Because IIRC the latter is usually useless in practice (too high), and afaik later recommendations by the EBU et al. recommend to clip to the mastering display volume.

> Why does libplacebo use 203.0 nits by default for max limit for SDR tonemapping instead of standard 100 nits?

This is the standard diffuse luminance level recommended by all later standards.

> Maybe it works better, but I have the doubt... Is it because it takes 0.2023 nits as minimum limit for SDR aswell? 203/0,2023 = ~ 100 units of dynamic range..

No, the other way around - we take 0.2023 nits as the SDR floor *because* that gives us a 1000:1 contrast.

**Reply by wyup** [reply]
> Because IIRC the latter is usually useless (too high) in practice

Well, MaxCLL can't be too high or useless, because MaxDML is usually higher and also higher than actual content. I understand the EBU rec. to clip to MaxMDL. It is true that MaxCLL sometimes is not the actual maximum luminance of the content, and it can be nearer MaxDML, but IMHO as result of video analysis MaxCLL represents about 99% max value in practice of the actual content. It is not worth to lose dynamic range by setting it to MaxDML because that is the mastering monitor display limit, not the limit of the actual content.
Usually MaxDML is either 1000, 4000 or 10000 for TV compatibility, too broad values to use for tonemapping limits. MaxCLL is the most relevant value of the content limit. I.e: a movie can be MaxDML of 1000 nits, and have 99% of the content with a 690 nits MaxCLL. We are losing about 310 nits of dynamic range (30%) unused for tonemapping that renders the picture darker when we are compressing 1000 to 200 nits.

> This is the standard diffuse luminance level recommended by all later standards.
> No, the other way around - we take 0.2023 nits as the SDR floor _because_ that gives us a 1000:1 contrast.

By curiosity:
Where is it derived that maxSDR/minSDR should be 1000:1 contrast ratio for SDR?
Where is it derived that 203 nits is the standard diffuse luminance level for SDR?

**Reply by haasn** [reply]
> Well, MaxCLL can't be too high or useless, because MaxDML is always higher and also higher than actual content.

Hmm, I could have sworn that there were samples I was testing on where MaxDML was 1000 but MaxCLL was close to 10,000 nits. But I can no longer find any. So maybe you're right, and we should default to MaxCLL instead?

> We are losing about 310 nits of dynamic range unused for tonemapping.

Keep in mind that these numbers are a bit misleading since we do tone mapping on PQ scale - so the difference between 1000 nits and 690 nits is only about 5% of headroom. Still wasteful, but not 1/3rd as it may seem.

> Where is SDR 1000:1 contrast value derived from? 

It is used as the default assumption for the average contrast of a typical LCD display. Actually, there have been multiple discussions about raising this contrast limit in light of the increasing popularity of OLED devices, so maybe we should reconsider that at this point in time?

> What's the difference between diffuse luminance level and contrast for choosing a standard SDR dynamic range?

I don't understand this question, can you rephrase?

> What default values does color_map_preset=high_quality change?

From the source code, it currently only enables contrast recovery:

```c
#define PL_COLOR_MAP_HQ_DEFAULTS                                \
    PL_COLOR_MAP_DEFAULTS                                       \
    .contrast_recovery      = 0.30f,
```

**Reply by wyup** [reply]
> Hmm, I could have sworn that there were samples I was testing on where MaxDML was 1000 but MaxCLL was close to 10,000 nits. But I can no longer find any. So maybe you're right, and we should default to MaxCLL instead?

Can you create or expose a new tone-mapping constant for manually setting a custom max_source_brightness? I would like to custom set this for Bt.2446a HDR to SDR tone-mapping instead of auto values (MaxDML at present from 'any' medatata source).

There is not a tone-mapping-param for this in mpv player either, but I can set custom libplacebo params. As an external example, rigaya's [QSVEnc](https://github.com/rigaya/QSVEnc/blob/master/QSVEncC_Options.en.md#--vpp-libplacebo-tonemapping-param1value1param2value2) utility implements all src_max, src_min, target_max, target_min values for libplacebo tonemapping.

To me, peak detection brightness is not an optimal solution for default tonemapping, it is too contrasty, restricts dynamic range and changes dinamically with picture into and in-between scenes. A whole tonemapping range is more natural and faithful to the content.

**Reply by wyup** [reply]
Any chance of a --src_max tonemapping parameter for custom max source brightness? I'd like to overrride mpv automatic choice of MaxDML for user selected range.

**Reply by kasper93** [reply]
@wyup: you can do it, for example `-vf=format=sig-peak=3`

**Reply by wyup** [reply]
> @wyup: you can do it, for example `-vf=format=sig-peak=3`

I don't mean a video filter, but to pass libplacebo src_max value for tonemapping up to a certain limit. We need this flexibility. Otherwise it's always maxDML, and that is often above actual content limit. Or just to maximise brightness and discard/clip the upper range when it's hardly present in the content. Avisynth and Vapoursinth interfaces for libplacebo implement these parameters.

**Reply by kasper93** [reply]
`--target-peak` controls maximum luminance of output signal.
`--vf=format=sig-peak=3` controls maxDML of source material.

**Reply by wyup** [reply]
ok, but that is " [#relative to the signal's reference white level](https://mpv.io/manual/master/#video-filters-%3Csig-peak%3E)". I'd like to (it makes more sense?) enter a simple nits value instead into libplacebo directly, or as a raw libplacebo option, without further input filtering.

[#--libplacebo-opts=src_max=<value>](https://mpv.io/manual/master/#options-libplacebo-opts)
The problem is that src_max is not defined on [https://libplacebo.org/options/](url), but it is on [avisynth libplacebo wrappers](http://www.avisynth.nl/index.php/Avslibplacebo#Tonemap).

**Reply by kasper93** [reply]
> ok, but that is " [#relative to the signal's reference white level](https://mpv.io/manual/master/#video-filters-%3Csig-peak%3E)". I'd like to (it makes more sense?) enter a simple nits value instead into libplacebo directly, or as a raw libplacebo option, without further input filtering.

it's just `nits / 203`. You we may add option to set directly nits value here, but for now you can divide it by 203 to get the value.

**Reply by wyup** [reply]
> `--target-peak` controls maximum luminance of output signal. `--vf=format=sig-peak=3` controls maxDML of source material.

`--target-peak` is a mpv only option, and we'd need `src_max` for ffmpeg and libplacebo.

Does ffmpeg's `--vf=format=sig-peak=3` filter really change maxDML from the file's HDR10 metadata that libplacebo reads for tonemapping?

**Reply by kasper93** [reply]
Yes, really really.

**Reply by wyup** [reply]
I'm not thinking of dynamic peak detection for tonemapping.
I want to disable dynamic peak detection and use bt2446a with data from:
`tone_map_metadata=<any|none|hdr10|hdr10plus|cie_y>`
How come a external filter such as `format` modifies engraved maxDML static parameter from the bitstream, if only hdr10 is available?
I can't find any `sig-peak` parameter to `format` video filter in ffmpeg-all and ffmpeg-filters documentation.

**Reply by kasper93** [reply]
https://mpv.io/manual/master/#video-filters-%3Csig-peak%3E

I told you three times how to do what you want. Is there anything I can do more?

**Reply by wyup** [reply]
sorry, I misunderstood `--vf=format=sig-peak=3` with a ffmpeg video filter.
As I said earlier, I need a src_max parameter for **libplacebo** so I can use it with **ffmpeg**.
I don't need mpv options at this time. This is not mpv discussion. This is libplacebo discussion, and as such, I'm asking for custom libplacebo parameters to include in ffmpeg custom libplacebo parameters.

**Reply by kasper93** [reply]
> This is not mpv discussion.

My bad this is different repository.

>  As I said earlier, I need a src_max parameter for libplacebo so I can use it with ffmpeg.

Set mastering metadata in your input file container, for mkv you can easily do that with `mkvpropedit`.

**Reply by wyup** [reply]
> Set mastering metadata in your input file container, for mkv you can easily do that with `mkvpropedit`

Thanks for the solution, but let me say I don't think it's elegant, because it conflicts with the stream's main mastering metadata.
I know I can set hdr10 metadata at mkv container level, but I don't use it because some media players ignore these, and I work with mp4 container mostly.
I think there should be a `tonemapping_param=src_max` in ffmpeg libplacebo filter that passes this value (and hopefully others like `src_min`, `dst_max`, `dst_min`) to libplacebo tonemapping function instead, directly from ffmpeg commandline.

This would be very useful as I could choose how many nits to set for SDR conversions, instead of default 203 nits, as in here:

http://www.avisynth.nl/index.php/Avslibplacebo#Tonemap
https://github.com/Asd-g/avslibplacebo?tab=readme-ov-file#tone-mapping
https://github.com/rigaya/QSVEnc/blob/master/QSVEncC_Options.en.md#--vpp-libplacebo-tonemapping-param1value1param2value2

"Use Avisynth then", you may say.. the problem with this avisynth (or vapoursynth) libplacebo interface is that its libplacebo dll library is outdated. For the moment QSVEnc offers exactly these features, but I'd like to use ffmpeg instead for flexibility, since it's the 'reference'. We really need the tonemapping customization.

**Reply by haasn** [reply]
@wyup what if we added extra options to `pl_options` to use purely as source/target metadata overrides, which the `pl_renderer` would then apply during the frame setup pass?

Then you could at least specify these overrides manually with every project that exposes the custom options string (e.g. `custom_options` in vf_libplacebo)

**Reply by kasper93** [reply]
>  I think there should be a tonemapping_param=src_max in ffmpeg libplacebo filter that passes this value (and hopefully others like src_min, dst_max, dst_min) to libplacebo tonemapping function instead, directly from ffmpeg commandline.

I think such information is not a part of libplacebo options interface and instead should be requested/fixed before calling libplacebo. 

For example if you are using ffmpeg CLI interface, this should be [hevc_metadata bsf](https://ffmpeg.org/ffmpeg-bitstream-filters.html#hevc_005fmetadata) filter, that inject your bitstream metadata. Which is currently not supported for mastering metadata, but it would be the place to inject/modify such info.

Alternatively like I said before, you can use container level metadata.

Alternatively you can use libplacebo API directly to inject custom side-data instead of the one attached to the AVFrame.

I don't think `pl_options` is the correct place to specify the input metadata. Especially if it is done selectively as you request. For example if `input mastering luminance` would be an option, that this mean `pl_options` should also include `input dimensions`, `input chroma location`, `input deinterlacing`, `input field order` and all other input frames metadata?

**Reply by haasn** [reply]
> Especially if it is done selectively as you request. For example if input mastering luminance would be an option, that this mean pl_options should also include input dimensions, input chroma location, input deinterlacing, input field order and all other input frames metadata?

That is my plan. I think we can support it pretty easily because `pl_renderer` already makes a copy of the `pl_frame` struct and applies overrides. If anything, I think this will make it easier for users to get predictable behavior, e.g. mpv could directly map the `--target-*` options to `pl_options`, and we could move the supporting logic there. (e.g. options like `target_contrast`, `target_gamut`, or even `dither_depth`).

**Reply by kasper93** [reply]
I don't see why libplacebo suddenly should include user application logic.

**Reply by haasn** [reply]
Well, in this case the benefit would mainly being able to share that logic between FFmpeg and libplacebo.

But maybe you're right, and we should just add this to `vf_libplacebo` directly.

**Reply by kasper93** [reply]
I see why you propose it. `pl_options` already maps to `pl_render_params`, not sure though it should really map to `pl_frame` and `pl_target_frame` with additional logic. It sound to me like something a caller should take care about, but if you feel like it. I would rather keep libplacebo scope limited.

**Reply by wyup** [reply]
Well, it looks libplacebo actually exposes those parameters since [QSVEnc](https://github.com/rigaya/QSVEnc/blob/master/QSVEncC_Options.en.md#--vpp-libplacebo-tonemapping-param1value1param2value2) already implements them. It's only a matter for ffmpeg's `vf_libplacebo` to add them via its `tonemapping_param`. You shouldn't touch the input source container tags or hevc_metadata bsf.

I don't know what you mean with `pl_options`.
